### ISSUE TITLE
Log files are not closed properly

### ISSUE THREAD

**Issue by 1dayac** [abishara/athena_meta#7]
Hello, Alex!
I found that log files are not closed when they should be. I have a restriction of 1000 open file handles and athena_meta inevitably crushes with error "Too many files are opened".
This sloppy quickfix worked for me (but I'm sure there is a better solution):
In log.py:

```
  def open_log(self):
    if self.log_file is None:
      self.log_file = open(self.log_path, "w")
```
to 

```
  def open_log(self):
    pass
```
and 

in  def log(self, message):

```
    self.log_file.write(message_str)
    self.log_file.flush()
```
to
```
   with open(self.log_path, "r+") as log.file:
       log_file.write(message_str)
       log_file.flush()
```

**Comment by abishara** [comment]
Hi,

Thank you for the suggestion.  Can I ask how many processes you specified?  An alternative solution would be to set the maximum to below your max file handle limit so that this should never be reached at once.  I’m a bit hesitant to accept your suggested changes because although this *can* alleviate failure from the restriction, it isn’t failsafe if all 1000 processes write at once.  Let me give it a bit more thought, but I’m happy you’ve found a temporary fix.

Also, not sure what version of Athena you are using, but we did some significant updates that are available on the main branch.  These updates introduced a dependence on a new fork of IDBA, referenced in the updated README, and fixed several bugs to significantly increase performance.  We found whole organisms that were unassembled with the previous version are now put together in <10 large contigs.  I hope you find it useful!  Sorry that there is not yet a tag released for these major revisions; this will come soon when we finalize fixes/updates.

Best!
alex

On Oct 25, 2017, at 11:51 AM, 1dayac <notifications@github.com<mailto:notifications@github.com>> wrote:


Hello, Alex!
I found that log files are not closed when they should be. I have a restriction of 1000 open file handles and athena_meta inevitably crushes with error "Too many files are opened".
This quickfix worked for me (but I'm there is a better solution):
In log.py:

def open_log(self):
if self.log_file is None:
self.log_file = open(self.log_path, "w")

to

def open_log(self):
pass

and

in def log(self, message):

self.log_file.write(message_str)
self.log_file.flush()


to

with open(self.log_path, "r+") as log.file:
self.log_file.write(message_str)
self.log_file.flush()

**Comment by 1dayac** [comment]
Thank you for your answer!
I observed this behaviour as with 1 and with 16 processes - so it is not because I used more processes than the number of possible opened file handles.
I'm using the latest version of Misha Kolmogorov's fork that uses ABruijn instead of Canu for OLC assembly.

**Comment by abishara** [comment]
Oh hmm, that is definitely a bug then and not desired.  Let me look a bit more into that when I get the chance to see if I can properly address it.  Thanks for pointing this out.

Great, that should contain most of the significant updates.  I will create a tag when we have a proper release version.

Thanks!
alex

On Oct 25, 2017, at 1:13 PM, 1dayac <notifications@github.com<mailto:notifications@github.com>> wrote:


Thank you for your answer!
I observed this behaviour as with 1 and with 16 processes - so it is not because I used more processes than the number of possible opened file handles.
I'm using the latest version of Misha Kolmogorov's fork that uses ABruijn instead of Canu for OLC assembly.
